### ISSUE TITLE
Logout?

### ISSUE THREAD

**Issue by nelsonic** [dwyl/auth_plug#11]
According to Suhail Doshi (co-founder of MixPanel), they did not have Logout functionality for the first 9 months because nobody _asked_ for it. People who didn't understand their product simply closed their browser and didn't return. That's the biggest problem in building an App. 
[![image](https://user-images.githubusercontent.com/194400/80181285-72a6ac80-85fc-11ea-90bc-a2dc35384330.png)](https://youtu.be/MABmQhOlmJA)
https://youtu.be/MABmQhOlmJA

![image](https://user-images.githubusercontent.com/194400/80181750-5ce5b700-85fd-11ea-95c7-bab871e8b9eb.png)
If people aren't _using_ the App, then Logout is the _least_ of their concerns.

The moment a **person _using_ our product** requests the ability to **logout** from the app (_e.g because they use a shared computer_) we will add an interface `AuthPlug.logout/1` where the single argument is the `conn` and it just returns the `conn` with the session deleted. 
We then need to figure out how to avoid session replay, because technically the JWT has not been invalidated it's just been deleted from the session cookies. 💭 

# _2021 Update_

I _still_ don't think that **`logout`** is something we (_or the people using our apps_) ***need***. 
But just for _testing_ purposes I think it will be _easier/faster_ to have the functionality. 

@iteles / @SimonLab what do you think?

# Todo

+ [x] Add `AuthPlug.logout/1` function that deletes the `session`.
+ [x] _Ideally_ we would also send this `logout` request back to the `Auth` app to prevent session fixation/hijacking ... but since the sessions are _stateless_ we would need to implement a whole other set of callback functions for checking session expiry/validity. 💭 `GOTO` Session Management: https://github.com/dwyl/auth/issues/30

**Comment by nelsonic** [comment]
While I've been testing the `logout/1` function I've discovered that we _cannot_ simply logout from the app that is using `auth_plug` we need to _also_ logout from the `auth` app ... 💭

**Comment by nelsonic** [comment]
Added the **`Logout`** button/link and `/logout` endpoint to the Example app: https://github.com/dwyl/auth_plug_example/pull/16
That _appears_ to work ...
![image](https://user-images.githubusercontent.com/194400/140432990-0dbc5ed1-3d1e-43ff-b010-69f86354faad.png)

When I click the `logout` link in the header the session cookie for the `localhost` is deleted:
![image](https://user-images.githubusercontent.com/194400/140433043-bd922ce5-320a-4d16-b3d1-fe2495938875.png)

_However_ ... since I'm still _authenticated_ with the **`auth`** app https://dwylauth.herokuapp.com
![image](https://user-images.githubusercontent.com/194400/140433124-70e0e1bb-43fe-44cd-b018-52b2639939df.png)

When I click on the **`admin`** link in the header of the ***Example*** App, 
it logs me back in without requiring any extra steps ... i.e. I'm not _truly_ logged out!

![image](https://user-images.githubusercontent.com/194400/140433246-163e2aff-ad5a-41d9-a2ba-6bb2085a0144.png)

i.e. because there is still a session cookie for `dwylauth.herokuapp.com` in my browser, 
I'm automatically logged back into the `auth_plug_example` app on my `localhost` without having to re-authenticate.

Believe it or not, this is actually a _feature_ for the **`auth`** App, not a _bug_!
We _want_ this because it means the Authenticated session for the "consumer" App is managed by **`auth`**. 

However from the perspective of implementing the **`logout`** functionality, 
it means that we need to have a `/logout` endpoint on the **`auth`** App 
that can be called by one of the "consumer" apps (_the ones using `auth_plug` ..._) 🤙 
So that when the person clicks the `logout` they are in fact logged out in _both_ places. 🚪 
Hopefully I'm making myself clear ... if not please just comment! 👍

**Comment by nelsonic** [comment]
Refresher: https://stackoverflow.com/questions/46626659/call-api-endpoint-with-elixir-phoenix

**Comment by SimonLab** [comment]
I was wondering why a session is created on the `auth` app while using the "consumer" app. This is done on `auth` with the following function:

```elixir
  def redirect_or_render(conn, person, state) do
    # check if we need to redirect to auth app or consumer app
    if is_nil(state) or state == "" do
      # No State > Display Welcome page on Auth site:
      conn
      |> AuthPlug.create_jwt_session(session_data(person))
      |> Auth.Log.info(%{status_id: 200, app_id: 1})
      |> render(:welcome, person: person, apps: App.list_apps(person.id))
    else
      # State > Redirect to consumer app:
      case get_client_secret_from_state(state) do
        0 ->
          unauthorized(conn, "invalid AUTH_API_KEY")

        secret ->
          conn
          |> AuthPlug.create_jwt_session(session_data(person))
          |> Auth.Log.info(%{status_id: 200, app_id: get_app_id(state)})
          |> redirect(external: add_jwt_url_param(person, state, secret))
      end
    end
  end
```

The last part is where the session is created:

```elixir
          # here we create the session in auth application
          |> AuthPlug.create_jwt_session(session_data(person))
          |> Auth.Log.info(%{status_id: 200, app_id: get_app_id(state)})
          # then we redirect back to the consumer app with the jwt in query parameter
          |> redirect(external: add_jwt_url_param(person, state, secret))
```

When we redirect to the application `auth_plug` will check if a jwt exists in the query parameter:

https://github.com/dwyl/auth_plug/blob/77963c86483c78acb3f2fe386416d67b528607e8/lib/auth_plug.ex#L29-L30

with the `get_jwt` function:
https://github.com/dwyl/auth_plug/blob/77963c86483c78acb3f2fe386416d67b528607e8/lib/token.ex#L133-L135

Then if the jwt is valid the session is created on the consumer app:
https://github.com/dwyl/auth_plug/blob/77963c86483c78acb3f2fe386416d67b528607e8/lib/auth_plug.ex#L34

So we have two sessions created, one in auth when login (with Github/Google or email) and the other one on the consumer app on redirection with a jwt in the query parameters.

I'm not sure why we need to have a session created on auth. We could easily avoid creating it and just return the jwt to the consumer app?

**Comment by nelsonic** [comment]
@SimonLab **_great_ question** and _thank you_ for diving deep enough into the code to understand it and ask! 🙌 

Yeah, my original reasoning for creating the `:jwt` session in _both_ the `auth` App and the "consumer" app via `auth_plug` was to streamline session renewal with the OAuth provider (e.g. GitHub or Google). 
But at the time we did not yet have Session management https://github.com/dwyl/auth/issues/30 ... 

While implementing **`Sessions`** last week https://github.com/dwyl/auth/pull/159 I realised that I don't _want_ to store the OAuth Refresh Token https://www.digitalocean.com/community/tutorials/an-introduction-to-oauth-2 in the `auth` DB ... 
because if the DB were ever compromised (_even though all records are encrypted_) it would be _very_ bad. 
So we will want to put the OAuth Refresh Token in the **`auth`** App's session ...

But, for now I think we can safely comment out the line in question to disable the **`auth`** App session creation. `# ...`
Later when we implement OAuth session renewal we can figure out what to do with it. 💭

**Comment by nelsonic** [comment]
We can remove the `:jwt` session:
```elixir
# Temporarily comment out jwt session. See: https://git.io/JXDzv
# |> AuthPlug.create_jwt_session(session_data(person))
```
But we'll need to update the tests...

Meanwhile the **`Session`** implementation requires a "callback" when we attempt to `logout` in the consumer app.

**Comment by nelsonic** [comment]
Implemented in #36 :shipit:

**Comment by nelsonic** [comment]
Closing as implemented. We can open a new issue if we need more features. 👍
